packages/optimizer/lib/transformers/AutoExtensionImporter.js

```diff
@@ -0,0 +1,269 @@
+ * using multiple strategies:
+ *
+ * - use validation rules to map used AMP tags to required AMP extensions.
+ * - use validation rules to map used AMP attributes to required AMP extensions.
+ * - manually specifiy attribute to extension mappings if this information is not available in the
+ *   validation rules.
+ * - mnullay implement AMP extension detection for a few corner cases.
+ *
+ * This importer also enables a shortcode `bindtext` instead of `data-amp-bind-text` for specifying
+ * AMP bindings when the square bracket notation (`[text]`) is not available. To avoid accidently
+ * rewriting non-AMP attributes, the transformer uses the AMP validation rules to only rename bindable
+ * attributes as specified in the validation rules.
+ */
+class AutoExtensionImporter {
+  constructor(config) {
+    this.enabled = config.autoExtensionImport !== 'false' && config.autoExtensionImport !== false;
+    this.log_ = config.log.tag('AutoExtensionImporter');
+
+    // We use the validation rules to infer extension imports. The rules are downloaded once and for
+    // efficency, we initially extract all needed rules
+    this.extensionSpec = validatorRules.fetch().then((rules) => {
+      // Map extension names to more info
+      const extensionsMap = new Map(rules.extensions.map((ext) => [ext.name, {
+        name: ext.name,
+        type: ext.extensionType === 'CUSTOM_TEMPLATE' ? 'custom-template' : 'custom-element',
+        version: ext.version.filter((v) => v !== 'latest'),
+      }]));
+      // Maps tags (e.g. amp-state) to their extension (e.g. amp-bind)
+      const tagToExtensionsMapping = new Map();
+      // Maps tags to their extension specific allowed attributes
+      // (e.g. amp-img => amp-fx => amp-fx-collection)
+      const tagToAttributeMapping = new Map();
+      // Maps tags to their bindable attributes (e.g. div => text)
+      const tagToBindAttributeMapping = new Map();
+
+      // Iterate over all available tags
+      for (const tag of rules.tags) {
+        const tagName = tag.tagName.toLowerCase();
+
+        // Map amp tags to their required extension(s)
+        if (tagName.startsWith('amp-')) {
+          tagToExtensionsMapping.set(
+              tagName,
+              tag.requiresExtension,
+          );
+        }
+        // Hack: fix missing attribute dependencies (e.g. amp-img => lightbox => amp-lightbox-gallery)
+        tag.attrs.filter((a) => manualAttributeToExtensionMapping.has(a.name))
+            .forEach((a) => {
+              a.requiresExtension = [manualAttributeToExtensionMapping.get(a.name)];
+            });
+
+        // Map attributes to tags and extensions (e.g. amp-img => amp-fx => amp-fx-collection)
+        tag.attrs.filter((a) => a.requiresExtension && a.requiresExtension.length > 0)
+            .forEach((a) => {
+              let attributeMapping = tagToAttributeMapping.get(tagName);
+              if (!attributeMapping) {
```

Comment:
This (ensuring an empty array exists as value) can be determined outside the loop rather than doing it in every iteration.

`tagToAttributeMapping.set(tagName, tagToAttributeMapping.get(tagName) || [])`